### PR TITLE
[Impeller] Support for image decode requests when the GPU sync switch is disabled

### PR DESCRIPTION
**Pull request by dnfield**
This feels a bit hacking (the `mutable` texture, doing the upload as soon as `impeller_texture` is requested with no guard, no tests), but it works locally to fix https://github.com/flutter/flutter/issues/124269

I'd like some feedback/sanity check before I go ahead and try to write tests. I'm also going to be unavailable for the rest of today, so if someone wants to move this forward sooner than later I'm fine with someone else taking over.

### CODE REVIEW

File: impeller/display_list/display_list_image_impeller_raster.cc [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+      return impeller::PixelFormat::kR8G8B8A8UNormInt;
+    case kRGBA_F16_SkColorType:
+      return impeller::PixelFormat::kR16G16B16A16Float;
+    case kBGR_101010x_XR_SkColorType:
+      return impeller::PixelFormat::kB10G10R10XR;
+    default:
+      return std::nullopt;
+  }
+  return std::nullopt;
+}
```
Comment (gaaclarke) [inline impeller/display_list/display_list_image_impeller_raster.cc:28]: This is copied from another file, can we share this logic?

File: impeller/display_list/display_list_image_impeller_raster.cc [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+      context->GetResourceAllocator()->CreateTexture(texture_descriptor);
+  if (!texture) {
+    FML_DLOG(ERROR) << "Could not create Impeller texture.";
+    return nullptr;
+  }
+
+  auto mapping = std::make_shared<fml::NonOwnedMapping>(
+      reinterpret_cast<const uint8_t*>(bitmap->getAddr(0, 0)),  // data
+      texture_descriptor.GetByteSizeOfBaseMipLevel(),           // size
+      [bitmap](auto, auto) mutable { bitmap.reset(); }          // proc
```
Comment (gaaclarke) [inline impeller/display_list/display_list_image_impeller_raster.cc:81]: nit: bitmap = std::move(bitmap)

File: lib/ui/painting/image_decoder_impeller.cc [hunk 3]
```diff
@@ -180,7 +181,9 @@ std::optional<DecompressResult> ImageDecoderImpeller::DecompressTexture(
 
   auto bitmap = std::make_shared<SkBitmap>();
   bitmap->setInfo(image_info);
-  auto bitmap_allocator = std::make_shared<ImpellerAllocator>(allocator);
+  auto bitmap_allocator = allocator
+                              ? std::make_shared<ImpellerAllocator>(allocator)
+                              : std::shared_ptr<ImpellerAllocator>(nullptr);
```
Comment (gaaclarke) [inline lib/ui/painting/image_decoder_impeller.cc:186]:
```suggestion
                              : nullptr;
```

File: lib/ui/painting/image_decoder_impeller.cc [hunk 4]
```diff
@@ -227,7 +233,9 @@ std::optional<DecompressResult> ImageDecoderImpeller::DecompressTexture(
   const auto scaled_image_info = image_info.makeDimensions(target_size);
 
   auto scaled_bitmap = std::make_shared<SkBitmap>();
-  auto scaled_allocator = std::make_shared<ImpellerAllocator>(allocator);
+  auto scaled_allocator = allocator
+                              ? std::make_shared<ImpellerAllocator>(allocator)
+                              : std::shared_ptr<ImpellerAllocator>(nullptr);
```
Comment (gaaclarke) [inline lib/ui/painting/image_decoder_impeller.cc:238]:
```suggestion
                              : nullptr;
```

File: impeller/display_list/display_list_image_impeller_raster.h [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+#include "flutter/fml/macros.h"
+#include "impeller/core/texture.h"
+#include "impeller/renderer/context.h"
+#include "third_party/skia/include/core/SkBitmap.h"
+
+namespace impeller {
+
+class AiksContext;
+
+class DlImageImpellerRaster final : public flutter::DlImage {
```
Comment (gaaclarke) [inline impeller/display_list/display_list_image_impeller_raster.h:17]: Add class docstring.

File: lib/ui/painting/image_decoder_impeller.cc [hunk 6]
```diff
@@ -424,6 +375,11 @@ void ImageDecoderImpeller::Decode(fml::RefPtr<ImageDescriptor> descriptor,
         auto upload_texture_and_invoke_result = [result, context,
                                                  bitmap_result =
                                                      bitmap_result.value()]() {
+          if (!bitmap_result.device_buffer) {
+            result(impeller::DlImageImpellerRaster::Make(
```
Comment (gaaclarke) [inline lib/ui/painting/image_decoder_impeller.cc:379]: This should have a comment explaining this branch.

File: impeller/display_list/display_list_image_impeller_raster.cc [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+
+// |DlImage|
+std::shared_ptr<impeller::Texture> DlImageImpellerRaster::impeller_texture()
+    const {
+  if (!texture_) {
+    auto context = context_.lock();
+    if (!context) {
+      return nullptr;
+    }
+    texture_ = UploadTextureToShared(context, bitmap_,
```
Comment (gaaclarke) [inline impeller/display_list/display_list_image_impeller_raster.cc:133]: So this is where the magic happens right?  Should this have a syncswitch too?

**Review by gaaclarke — CHANGES_REQUESTED** [review]
This needs a test =)

**Comment by gaaclarke** [comment]
Sorry, I just read your description.

Did we verify that we can't actually create the texture with metal while backgrounded?  Like make both branches of the sync switch upload the texture?  It is possible that certain calls, like making textures, may just work in metal with shared vram / ram.

Is the idea that all the code that would access the texture would already be guarded by the sync_switch higher up in the stack trace?  Can we add an assert to make sure that's the case?

File: impeller/display_list/display_list_image_impeller_raster.cc (hunk 1 above)
Comment (dnfield) [reply impeller/display_list/display_list_image_impeller_raster.cc:28]: Yeah, I meant to make this shared.

**Comment by dnfield** [comment]
> Sorry, I just read your description.
> 
> Did we verify that we can't actually create the texture with metal while backgrounded? Like make both branches of the sync switch upload the texture? It is possible that certain calls, like making textures, may just work in metal with shared vram / ram.
> 
> (I guess even if that works we still need to account for opengl in impeller where it doesn't work.)
> 
> Is the idea that all the code that would access the texture would already be guarded by the sync_switch higher up in the stack trace? Can we add an assert to make sure that's the case?

I'm not really clear on this. We're not allowed to encode commands. A shared texture doesn't require any command encoding from what I can tell. So maybe the shared path is just ok as is?

**Comment by dnfield** [comment]
There's only command buffer creation/usage if mipmapping is created, which I think is fine to skip for this case. I'll try an alternative patch for that, which should be a bit simpler.

**Comment by dnfield** [comment]
I think https://github.com/flutter/engine/pull/40963 is a better appraoch.

**Pull request closed without merging**